packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx

```diff
@@ -362,42 +366,49 @@ export const useGridFilter = (
 
   const flatFilteringMethod = React.useCallback<GridStrategyProcessor<'filtering'>>(
     (params) => {
-      if (props.filterMode === 'client' && params.isRowMatchingFilters) {
-        const tree = gridRowTreeSelector(apiRef);
-        const rowIds = (tree[GRID_ROOT_GROUP_ID] as GridGroupNode).children;
-        const filteredRowsLookup: Record<GridRowId, boolean> = {};
-        for (let i = 0; i < rowIds.length; i += 1) {
-          const rowId = rowIds[i];
-          let isRowPassing;
-          if (typeof rowId === 'string' && rowId.startsWith('auto-generated-group-footer')) {
-            isRowPassing = true;
-          } else {
-            const { passingFilterItems, passingQuickFilterValues } =
-              params.isRowMatchingFilters(rowId);
-            isRowPassing = passFilterLogic(
-              [passingFilterItems],
-              [passingQuickFilterValues],
-              params.filterModel,
-              apiRef,
-            );
-          }
-          filteredRowsLookup[rowId] = isRowPassing;
-        }
+      if (props.filterMode !== 'client' || !params.isRowMatchingFilters) {
         return {
-          filteredRowsLookup,
-          // For flat tree, the `visibleRowsLookup` and the `filteredRowsLookup` are equals since no row is collapsed.
-          visibleRowsLookup: filteredRowsLookup,
+          visibleRowsLookup: new Set(),
+          filteredRowsLookup: new Set(),
           filteredDescendantCountLookup: {},
         };
       }
 
+      const rows = props.rows;
+      const filteredRowsLookup = new Set<GridRowId>();
+      const { isRowMatchingFilters } = params;
+
+      const result = {
+        passingFilterItems: null,
+        passingQuickFilterValues: null,
+      };
+
+      for (let i = 0; i < rows.length; i += 1) {
+        const row = rows[i];
+
+        isRowMatchingFilters(row, undefined, result);
+
+        const isRowPassing = passFilterLogicSingle(
+          result.passingFilterItems,
+          result.passingQuickFilterValues,
+          params.filterModel,
+          apiRef,
+        );
+
+        if (isRowPassing) filteredRowsLookup.add(row.id);
+      }
+
+      // XXX: Is props.rows what we want?
+      // XXX: Handle footer rows
```

Comment:
TODO: Add autogenerated rows after the loop.

Reply:
Why not using `(tree[GRID_ROOT_GROUP_ID]).children;` here as before?

Reply:
See [point 9](https://github.com/mui/mui-x/pull/9120#issuecomment-1566215128)

Reply:
This quick benchmark shows that at least property access should be faster when using `Map`: https://jsperf.app/petawa

Reply:
I benchmarked by replacing our state model by a `Map` inside d8. The complexity of JS engines makes it hard to predict the performance solely with a microbenchmark, we should benchmark further the whole process before switching. I'll start with the other points, once I'm done I can look further into using a `Map`.